backend/src/org/commcare/suite/model/OfflineUserRestore.java

```diff
@@ -0,0 +1,135 @@
+import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+
+/**
+ * User restore xml file sometimes present in apps.
+ * Used for offline (demo user) logins.
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ * @author Aliza Stone (astone@dimagi.com)
+ */
+public class OfflineUserRestore implements Persistable {
+    public static final String STORAGE_KEY = "OfflineUserRestore";
+    private int recordId = -1;
+    private String restore;
+    private String reference;
+    private String username;
+    private String password;
+
+    public OfflineUserRestore() {
+    }
+
+    public OfflineUserRestore(String reference) {
+        this.reference = reference;
+        this.username = parseUsernameFromRestoreFile(reference);
+        this.password = PropertyUtils.genUUID();
+    }
+
+    public static OfflineUserRestore buildInMemoryUserRestore(InputStream restoreStream) throws IOException {
+        OfflineUserRestore offlineUserRestore = new OfflineUserRestore();
+        offlineUserRestore.restore = new String(StreamsUtil.inputStreamToByteArray(restoreStream));
+        return offlineUserRestore;
+    }
+
+    public InputStream getRestoreStream() {
+        if (reference == null) {
+            return getInMemoryStream();
+        } else {
+            return getStreamFromReference();
+        }
+    }
+
+    private InputStream getStreamFromReference() {
+        try {
+            Reference local = ReferenceManager._().DeriveReference(reference);
+            return local.getStream();
+        } catch (IOException | InvalidReferenceException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private InputStream getInMemoryStream() {
+        try {
+            return new ByteArrayInputStream(restore.getBytes("UTF-8"));
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public String getReference() {
+        return reference;
+    }
+
+    public String getUsername() {
+        return username;
+    }
+
+    public String getPassword() {
+        return password;
+    }
+
+    @Override
+    public void readExternal(DataInputStream in, PrototypeFactory pf)
+            throws IOException, DeserializationException {
+        this.recordId = ExtUtil.readInt(in);
+        this.reference = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+        this.restore = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+    }
+
+    @Override
+    public void writeExternal(DataOutputStream out) throws IOException {
+        ExtUtil.writeNumeric(out, recordId);
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(reference));
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(restore));
+    }
+
+    @Override
+    public void setID(int ID) {
+        recordId = ID;
+    }
+
+    @Override
+    public int getID() {
+        return recordId;
+    }
+
+    private static String parseUsernameFromRestoreFile(String localRestoreReference) {
```

Comment:
Is there a reason we don't use the normal user parser and parser factory here? It seems like that re-uses more stable code rather than redoing the parsing on its own.